docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+to another node, eg: when the node goes low on resources.
+
+This can be thought as an infrastructure issue independent to the s3gw.
+In this case, the *One Active pipeline* model fully restores the service by
+rescheduling a new POD somewhere in the cluster.
+
+### Cluster's node failure
+
+This case is when the `radosgw` process stops due to a cluster's node failure.
+The *One Active pipeline* model fully restores the service by
+rescheduling a new POD somewhere in the cluster.
+
+### radosgw's failure due to a bug not related to a certain input pattern
+
+This case is when the `radosgw` process crashes due to a bug not directly
+related to any input type.
+
+Examples:
+
+- Memory leaks
+- Memory corruptions (stack or heap corruptions)
+- Periodic operations or routines not related to an input (GC, measures,
+  telemetry, etc)
+
+For a bug like this, the *One Active pipeline* model could guarantee
+the user's operations *until the next occurrence* of the same malfunctioning.
+
+A definitive solution would be available only when a patch for the issue
+has released.
+
+### radosgw's failure due to a bug related to a certain input pattern
+
+This case is when the `radosgw` process crashes due to a bug directly
+related to a certain input type.
+
+Examples:
+
+- Putting Buckets with some name's pattern
+- Putting Objects that have a certain size
+- Performing an admin operation over a suspended user
+
+For a bug like this, the *One Active pipeline* model could guarantee
+the user's operations under the condition that the crash-triggering
+input is recognized by the user and thus its submission blocked.
+
+### PV Data corruption at application level due to radosgw's bug
+
+This case is when the state on the `PV` corrupts due to a `radosgw`
+bug.
+
+In this unfortunate scenario, the *One Active pipeline* can hardly help.
+A restart could evenly fix the problem or trigger an endless restarting loop.
+
+The fix for this could contemplate an human intervention.
+A definitive solution would be available only when a patch for the issue
+is available.
+
+### PV Data corruption at application level due to radosgw's anomalous exit
+
+This case is when the state on the PV corrupts due to a `radosgw`
+anomalous exit, eg: after a node failure.
+
+This scenario could even be NOT possible because of the safety features
+implemented on the radosgw SFS backend.
+
+The fix for this could contemplate an human intervention.
+
+## Measuring s3gw failures on Kubernetes
+
+After reviewing the cases, we can say that what can be actually solved with
+an HA model with s3gw is when the failure is not dependent to applicative bugs.
+We can handle temporary issues involving the infrastructure that is hosting
+the `radosgw` process.
+
+So, we are mostly interested in observing and measuring the timing of
+Kubernetes when a `radosgw`'s POD is being rescheduled.
+
+Hence, The idea to collect measures regarding a series of restarts
```

Comment:
I think we need separate things.

The first is the (kill - re)start loop timing outside of k8s/LH (so we have a baseline and can measure of what k8s adds, and how slow we are when exiting cleanly, crashing with no ops in flight, vs a bazillion point three five parallel ops on-going).

Then, fault detection times for k8s - how long until it notices that the process has crashed (that should be quick due to the generated signal), but what about the process *hanging*? (e.g., crashes are separate from timeouts)

For node failures, again, we need to understand which factors affect k8s detecting that and reacting to them, and what latencies are introduced by k8s/LH.

Actively asking k8s to restart is different (see case one, switch- vs fail-over). That should be smooth, but is not actually a failure scenario. Probably worth handling in a separate section.